test/test_type_promotion.py

```diff
@@ -0,0 +1,225 @@
+
+                    func = getattr(torch, f)
+                    func(x, y).sum().backward()
+
+    # verifies that a.add(b) is the same as a.to(b.dtype).add(b) in cases
+    # where that should hold.
+    def test_many_promotions(self):
+        from_to = {
+            torch.float16: torch.float32,
+            torch.half: torch.float16,
+            torch.int: torch.long,
+            torch.uint8: torch.long,
+            torch.uint8: torch.float,
+            torch.int: torch.float,
+            torch.int: torch.double,
+            torch.int16: torch.long,
+            torch.float16: torch.double,
+            torch.bool: torch.long,
+            torch.bool: torch.float
+        }
+
+        for k, v in from_to.items():
+            a = torch.rand([3, 3], device=self.device).to(k)  # no _th_uniform for half on cpu.
+            b = torch.rand([3, 3], device=self.device).to(v)
+            c = a.add(b)
+            d = a.to(v).add(b)
+            self.assertEqual(c.dtype, d.dtype, message='from {} to {}'.format(k, v))
+            self.assertEqual(c, d, message='from {} to {}'.format(k, v))
+
+    def test_non_promoting_ops(self):
+        x = torch.ones(4, dtype=torch.double)
+        err = 'expected dtype .ouble .*but got dtype .loat'
+        self.assertRaisesRegex(RuntimeError, err,
+                               lambda: torch.neg(torch.ones(4, dtype=torch.float), out=x))
+        self.assertRaisesRegex(RuntimeError, err,
+                               lambda: torch.lerp(x, torch.ones(4, dtype=torch.float), 1))
+
+    def test_alpha_mismatch(self):
+        x = torch.ones(4, dtype=torch.int)
+        err = 'alpha must not be'
+        self.assertRaisesRegex(RuntimeError, err,
+                               lambda: torch.add(x, x, alpha=1.1))
+
+    def test_booleans(self):
+        onedim = torch.tensor([True])
+
+        self.assertEqual(onedim + onedim, onedim)
+        self.assertEqual(onedim + True, onedim)
+        self.assertEqual(torch.add(True, True), True)
+        self.assertEqual(torch.add(False, False), False)
+        self.assertEqual(torch.add(False, True), True)
+
+    def test_create_bool_tensors(self):
+        self.assertEqual(torch.arange(False, True), torch.tensor([0], dtype=torch.int64))
+        self.assertEqual(torch.arange(True), torch.tensor([0], dtype=torch.int64))
+        self.assertEqual(torch.arange(False, True, 0.5), torch.tensor([0, 0.5], dtype=torch.get_default_dtype()))
+        self.assertEqual(torch.arange(False, False), torch.ones(0, dtype=torch.int64))
+
+        self.assertEqual(torch.linspace(False, True), torch.linspace(0, 1))
+        self.assertEqual(torch.logspace(False, True), torch.logspace(0, 1))
+
+        # this seems like odd behavior but ints also create float tensors, numpy doesn't have this function.
+        self.assertEqual(torch.scalar_tensor(False), torch.tensor(0.))
+
+@unittest.skipIf(not torch.cuda.is_available(), "no cuda")
+class TestTypePromotionCuda(TestTypePromotion):
```

Comment:
sure, but why do you want to do it at the class level?

Reply:
all this class does is set `self.device = 'cuda'`, it's just a helper to run all the tests with cuda instead of cpu.

Normally the @cudaonly method level filter would be for tests that don't make sense on CPU, e.g., if it was testing `Half` operations that are only supported on cuda, or when the particular test implementation explicitly sets the device.